### ISSUE TITLE
How to gen same model with same parameter?

### ISSUE THREAD

**Issue by baibaiw5** [facebookresearch/fastText#257]
When i run the following cmd a lot of time,the gen model is different.How can I gen same model with same parameter?

fasttext supervised -input ./fasttext-20170502.txt -output model  -dim 100 -lr 0.1 -wordNgrams 1 -minCount 5 -bucket 10000000 -epoch 50 -thread 1

**Comment by spate141** [comment]
try using more threads, if your system supports more, `-thread 4`
try different loss, `-loss hs`

**Comment by cpuhrsch** [comment]
Hey @baibaiw5,

The initial parameters for the model are chosen randomly. This implies that the resulting model will converge to different values. You can find this here

`https://github.com/facebookresearch/fastText/blob/master/src/fasttext.cc#L610`

More specifically

```
    input_ = std::make_shared<Matrix>(dict_->nwords()+args_->bucket, args_->dim);
    input_->uniform(1.0 / args_->dim);
```

You'd have to change this to get the exact same parameters after training (and every other instance of randomization, such as negative sampling, etc.). You also can't use more than one thread, since the underlying algorithm (hogwild) does not care about thread-safety for updates. Having said that, you should not see a large variance in performance regardless. What makes you worry about reproducibility at this level?

Thanks,
Christian

**Comment by baibaiw5** [comment]
@cpuhrsch 
Thank you for your comments. I will close this issue.

**Comment by arnaudsj** [comment]
@cpuhrsch I know the issue is closed, but I am really curious if there is no desire at all to make a run reproduceable. Specially in the case of trying to optimize the hyper-parameters, it is quite important to use the same seed.

So far I have been able to mitigate the randomness of the model training by using cross validation when performing parameter tuning, but this does increase the run time as a result of course.

**Comment by atomkirk** [comment]
Also makes testing hard. I have some tests that pass 90% of the time but every once in a while it makes a wrong prediction. The training data is static and same every time so it seems related to this randomness within fasttext itself.
